builtin/provisioners/file/resource_provisioner.go

```diff
@@ -48,13 +43,52 @@ func (p *ResourceProvisioner) Apply(
 
 // Validate checks if the required arguments are configured
 func (p *ResourceProvisioner) Validate(c *terraform.ResourceConfig) (ws []string, es []error) {
-	v := &config.Validator{
-		Required: []string{
-			"source",
-			"destination",
-		},
+	numDst := 0
+	numSrc := 0
+	for name := range c.Raw {
+		switch name {
+		case "destination":
+			numDst++
+		case "source", "content":
+			numSrc++
+		default:
+			es = append(es, fmt.Errorf("Unknown configuration '%s'", name))
+		}
+	}
+	if numSrc != 1 || numDst != 1 {
+		es = append(es, fmt.Errorf("Must provide one  of 'content' or 'source' and 'destination' to file"))
+	}
+	return
+}
+
+// getSrc returns the file to use as source
+func (p *ResourceProvisioner) getSrc(c *terraform.ResourceConfig) (string, error) {
+	var src string
+
+	sRaw, ok := c.Config["source"]
+	if ok {
+		if src, ok = sRaw.(string); !ok {
+			return "", fmt.Errorf("Unsupported 'source' type! Must be string.")
+		}
+	}
+
+	content, ok := c.Config["content"]
+	if ok {
+		file, err := ioutil.TempFile("", "tf-file-content")
```

Comment:
Just my opinion, i think the few extra lines of code outweigh the round trip to the disk and the orphaned temp file left on the disk. The code in the gist works, you can refactor it in and make it your own if you agree with the approach. Anyway, great feature and elegant interface ( adding the content parameter ), will definitely use it my self.